updates_and_signals/atomic_message_handlers/workflow.py

```diff
@@ -0,0 +1,198 @@
+        self.state.cluster_started = True
+        self.state.nodes = {str(k): None for k in range(25)}
+        workflow.logger.info("Cluster started")
+
+    @workflow.signal
+    async def shutdown_cluster(self):
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        self.state.cluster_shutdown = True
+        workflow.logger.info("Cluster shut down")
+
+    @workflow.update
+    async def allocate_n_nodes_to_job(
+        self,
+        task_name: str,
+        num_nodes: int,
+    ) -> List[str]:
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        if self.state.cluster_shutdown:
+            raise RuntimeError(
+                "Cannot allocate nodes to a job: Cluster is already shut down"
+            )
+
+        async with self.nodes_lock:
+            unassigned_nodes = [k for k, v in self.state.nodes.items() if v is None]
+            if len(unassigned_nodes) < num_nodes:
+                raise ValueError(
+                    f"Cannot allocate {num_nodes} nodes; have only {len(unassigned_nodes)} available"
+                )
+            assigned_nodes = unassigned_nodes[:num_nodes]
+            # This await would be dangerous without nodes_lock because it yields control and allows interleaving.
+            await self._allocate_nodes_to_job(assigned_nodes, task_name)
+            self.state.max_assigned_nodes = max(
+                self.state.max_assigned_nodes,
+                len([k for k, v in self.state.nodes.items() if v is not None]),
+            )
+            return assigned_nodes
+
+    async def _allocate_nodes_to_job(self, assigned_nodes: List[str], task_name: str):
+        await workflow.execute_activity(
+            allocate_nodes_to_job,
+            AllocateNodesToJobInput(nodes=assigned_nodes, task_name=task_name),
+            start_to_close_timeout=timedelta(seconds=10),
+        )
+        for node in assigned_nodes:
+            self.state.nodes[node] = task_name
+
+    @workflow.update
+    async def delete_job(self, task_name: str) -> str:
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        assert not self.state.cluster_shutdown
+        async with self.nodes_lock:
+            nodes_to_free = [k for k, v in self.state.nodes.items() if v == task_name]
+            # This await would be dangerous without nodes_lock because it yields control and allows interleaving.
+            await self._deallocate_nodes_for_job(nodes_to_free, task_name)
+        return "Done"
+
+    async def _deallocate_nodes_for_job(self, nodes_to_free: List[str], task_name: str):
+        await workflow.execute_activity(
+            deallocate_nodes_for_job,
+            DeallocateNodesForJobInput(nodes=nodes_to_free, task_name=task_name),
+            start_to_close_timeout=timedelta(seconds=10),
+        )
+        for node in nodes_to_free:
+            self.state.nodes[node] = None
+
+    def get_assigned_nodes(self) -> List[str]:
```

Comment:
Any concerns that this list may be mutated during iteration by a running update since it is not called under lock from `run`?